src/dotnet/commands/dotnet-tool/restore/ToolRestoreCommand.cs

```diff
@@ -0,0 +1,197 @@
+            IToolManifestFinder toolManifestFinder = null,
+            ILocalToolsResolverCache localToolsResolverCache = null,
+            DirectoryPath? nugetGlobalPackagesFolder = null,
+            IReporter reporter = null)
+            : base(result)
+        {
+            _options = appliedCommand ?? throw new ArgumentNullException(nameof(appliedCommand));
+
+            if (toolPackageInstaller == null)
+            {
+                (IToolPackageStore,
+                    IToolPackageStoreQuery,
+                    IToolPackageInstaller installer) toolPackageStoresAndInstaller
+                        = ToolPackageFactory.CreateToolPackageStoresAndInstaller(
+                            additionalRestoreArguments: appliedCommand.OptionValuesToBeForwarded());
+                _toolPackageInstaller = toolPackageStoresAndInstaller.installer;
+            }
+            else
+            {
+                _toolPackageInstaller = toolPackageInstaller;
+            }
+
+            _toolManifestFinder = toolManifestFinder
+                                  ?? new ToolManifestFinder(new DirectoryPath(Directory.GetCurrentDirectory()));
+
+            _localToolsResolverCache = localToolsResolverCache ??
+                                       new LocalToolsResolverCache(
+                                           new FileSystemWrapper(),
+                                           new DirectoryPath(Path.Combine(CliFolderPathCalculator.ToolsResolverCachePath)),
+                                           _localToolResolverCacheVersion);
+
+            _nugetGlobalPackagesFolder =
+                nugetGlobalPackagesFolder ?? new DirectoryPath(NuGetGlobalPackagesFolder.GetLocation());
+            _reporter = reporter ?? Reporter.Output;
+            _errorReporter = reporter ?? Reporter.Error;
+
+            _configFilePath = appliedCommand.ValueOrDefault<string>("configfile");
+            _source = appliedCommand.ValueOrDefault<string[]>("add-source");
+            _verbosity = appliedCommand.SingleArgumentOrDefault("verbosity");
+        }
+
+        public override int Execute()
+        {
+            FilePath? customManifestFileLocation = GetCustomManifestFileLocation();
+
+            FilePath? configFile = null;
+            if (_configFilePath != null) configFile = new FilePath(_configFilePath);
+
+            IReadOnlyCollection<ToolManifestFindingResultSinglePackage> packagesFromManifest =
+                _toolManifestFinder.Find(customManifestFileLocation);
+
+            Dictionary<RestoredCommandIdentifier, RestoredCommand> dictionary =
+                new Dictionary<RestoredCommandIdentifier, RestoredCommand>();
+
+            Dictionary<PackageId, ToolPackageException> toolPackageExceptions =
+                new Dictionary<PackageId, ToolPackageException>();
+
+            foreach (var package in packagesFromManifest)
+            {
+                string targetFramework =
```

Comment:
manifest file could break due to different versions of SDK. for example, there is no runtime 3.0 version available. pin to a lower version can mitigate this issue

Reply:
and I do plan to surface warning when manifest version is higher than bundled. it is in the pending issues

Reply:
I'm not following how it would break.  If the target framework for restore is netcoreapp3.0, then it should successfully restore a tool that only has a netcoreapp2.1 version, for example.